backend-rust/src/graphql_api.rs

```diff
@@ -1812,8 +1818,93 @@ impl Contract {
         todo_api!()
     }
 
-    async fn tokens(&self, skip: u32, take: u32) -> ApiResult<TokensCollectionSegment> {
-        todo_api!()
+    // This function fetches CIS2 tokens associated to a given contract, ordered by
+    // their creation index in descending order. It retrieves the most recent
+    // tokens first, with support for pagination through `skip` and `take`
+    // parameters.
+    // - `skip` determines how many of the most recent tokens to skip.
+    // - `take` controls the number of tokens to return, respecting the collection
+    //   limit.
+    async fn tokens(
+        &self,
+        ctx: &Context<'_>,
+        skip: Option<u64>,
+        take: Option<u64>,
+    ) -> ApiResult<TokensCollectionSegment> {
+        let config = get_config(ctx)?;
+        let pool = get_pool(ctx)?;
+
+        let max_token_index = sqlx::query!(
+            "SELECT MAX(token_index_per_contract) AS max_token_index
+            FROM tokens
+            WHERE tokens.contract_index = $1 AND tokens.contract_sub_index = $2",
+            self.contract_address_index.0 as i64,
+            self.contract_address_sub_index.0 as i64,
+        )
+        .fetch_one(pool)
+        .await?
+        .max_token_index
+        .unwrap_or(0) as u64;
+
+        let max_index = max_token_index.saturating_sub(skip.unwrap_or(0));
+        let limit = i64::try_from(take.map_or(config.contract_tokens_collection_limit, |t| {
+            config.contract_tokens_collection_limit.min(t)
+        }))?;
+
+        let tokens = sqlx::query!(
+            "SELECT
+                total_supply as raw_total_supply,
+                token_index_per_contract,
+                token_id,
+                contract_index,
+                contract_sub_index,
+                token_address,
+                metadata_url,
+                init_transaction_index
+            FROM tokens
+            WHERE tokens.contract_index = $1 AND tokens.contract_sub_index = $2
+                AND tokens.token_index_per_contract <= $3
+            ORDER BY tokens.token_index_per_contract DESC
+            LIMIT $4
+            ",
+            self.contract_address_index.0 as i64,
+            self.contract_address_sub_index.0 as i64,
+            max_index as i64,
+            limit as i64 + 1,
+        )
+        .fetch_all(pool)
+        .await?;
+
+        let mut items: Vec<Token> = tokens
+            .into_iter()
+            .map(|row| Token {
+                raw_total_supply:       row.raw_total_supply,
+                token_id:               row.token_id,
+                contract_index:         row.contract_index,
+                contract_sub_index:     row.contract_sub_index,
+                token_address:          row.token_address,
+                metadata_url:           row.metadata_url,
+                init_transaction_index: row.init_transaction_index,
+            })
+            .collect();
```

Comment:
I think we could use `query_as!` here
```suggestion
        let items = sqlx::query_as!(Token,
            "SELECT
                total_supply as raw_total_supply,
                token_index_per_contract,
                token_id,
                contract_index,
                contract_sub_index,
                token_address,
                metadata_url,
                init_transaction_index
            FROM tokens
            WHERE tokens.contract_index = $1 AND tokens.contract_sub_index = $2
                AND tokens.token_index_per_contract <= $3
            ORDER BY tokens.token_index_per_contract DESC
            LIMIT $4
            ",
            self.contract_address_index.0 as i64,
            self.contract_address_sub_index.0 as i64,
            max_index as i64,
            limit as i64 + 1,
        )
        .fetch_all(pool)
        .await?;
```

Reply:
`Token` has no `token_index_per_contract` field. I thought about adding it but it is `dead_code` in the `Token` struct and not really used for anything (wasn't happy with that).

Reply:
Can't we just admit it from the SELECT ?
```suggestion
        let items = sqlx::query_as!(Token,
            "SELECT
                total_supply as raw_total_supply,
                token_id,
                contract_index,
                contract_sub_index,
                token_address,
                metadata_url,
                init_transaction_index
            FROM tokens
            WHERE tokens.contract_index = $1 AND tokens.contract_sub_index = $2
                AND tokens.token_index_per_contract <= $3
            ORDER BY tokens.token_index_per_contract DESC
            LIMIT $4
            ",
            self.contract_address_index.0 as i64,
            self.contract_address_sub_index.0 as i64,
            max_index as i64,
            limit as i64 + 1,
        )
        .fetch_all(pool)
        .await?;
```